scd/usbtoken.c

```diff
@@ -0,0 +1,737 @@
+ *	as returnet from CT_list()
+ * @param lr actual length of @param readers
+ * @param serial the serial number of the usb smartcard reader
+ * @param port the port reader with the given serial or NULL
+ * @return 0 if success
+ */
+int
+token_filter_by_serial(const unsigned char *readers, const unsigned short lr, const char *serial,
+		       unsigned short *port)
+{
+	ASSERT(readers);
+	ASSERT(serial);
+
+	unsigned char *po;
+	unsigned short idx;
+	unsigned short iport;
+
+	char *s = (char *)mem_memcpy(readers, lr);
+	/* readers: |-|-|---15---|---X---|-|-|
+	 *	 fields  1 2     3       4    5 6
+	 * 		1: uint8_t libusb_get_bus_number()
+	 * 		2: uint8_t libusb_get_device_address
+	 * 		3: string  "SmartCard-HSM ("
+	 * 		4: string  iSerialNumber
+	 * 		5: string  ")"
+	 * 		6: NULL byte to delimit next reader
+	 */
+	po = (unsigned char *)readers;
+	idx = 0;
+
+	while (idx < lr) {
+		iport = *po << 8 | *(po + 1);
+		po += 2;
+		idx += 2;
+
+		po += 15;
+
+		po[lr - 15 - 1 - 1 - 1 - 1] = 0;
+
+		size_t s_len = strlen((const char *)po);
+
+		if ((s_len > 0) && (strncmp((const char *)po, serial, s_len) == 0)) {
+			TRACE("USBTOKEN: token_filter_by_serial() serial: %s\nport %d", serial,
+			      iport);
+			*port = iport;
+			TRACE("port: 0x%04x", *port);
+			mem_free(s);
+			return 0;
+		}
+
+		po += 15 + s_len + 1;
+		idx += 15 + s_len + 1;
+	}
+
+	mem_free(s);
+	return -1;
+}
+
+/**
+ * initializes the ctapi interface to the token reader.
+ * If the reader is unplugged during runtime this will ensure it is either
+ * reaquired using the cirrect usb port or not at all
+ */
+static int
+usbtoken_init_ctapi_int(usbtoken_t *token)
+{
+	TRACE("USBTOKEN: usbtoken_init_ctapi_int");
+
+	ASSERT(token);
+
+	int rc;
+	unsigned short lr, port;
+	unsigned char *readers = mem_alloc0(MAX_CT_READERS_SIZE);
+	if (!readers) {
+		ERROR("Failed to allocate memoryfor ct_reader");
+		mem_free(readers);
+		return -1;
+	}
+
+	lr = MAX_CT_READERS_SIZE;
+	CT_list(readers, &lr, 0);
+
+	if (lr <= 0) {
+		ERROR("No usb token reader found.");
+		return -1;
+	}
+
+	rc = token_filter_by_serial(readers, lr, token->serial, &port);
+	if (rc != 0) {
+		ERROR("Could not find specified token reader with serial %s", token->serial);
+		return -1;
+	}
+	token->port = port;
+
+	rc = CT_init(token->ctn, port);
+	if (rc != 0) {
+		ERROR("USBTOKEN: Token reader initialization failed. Ret code: %d", rc);
+		return -1;
+	}
+
+	requestICC(token->ctn);
+
+	rc = queryPIN(token->ctn);
+	if (rc != 0x9000) {
+		selectHSM(token->ctn);
+		rc = queryPIN(token->ctn);
+		TRACE("usbtoken_init queryPIN: %d", rc);
+	}
+
+	return 0;
+}
+
+/**
+ * Initializes a usb token
+ */
+usbtoken_t *
+usbtoken_init(const char *serial)
+{
+	TRACE("USBTOKEN: usbtoken_init");
+
+	ASSERT(serial);
+
+	int rc;
+	usbtoken_t *token = NULL;
+
+	token = mem_new0(usbtoken_t, 1);
+	ASSERT(token);
+
+	token->locked = true;
+	token->ctn = g_ctn++;
+	token->serial = mem_strdup(serial);
```

Comment:
allocation should be done in a _new function
serial is strdup'ed below, so delete this here